custom_model_runner/datarobot_drum/drum/drum.py

```diff
@@ -759,26 +775,65 @@ def _run_inside_docker(self, options, run_mode, raw_arguments):
 
     def _maybe_build_image(self, docker_image_or_directory):
         ret_docker_image = None
-        client = docker.client.from_env()
 
         if os.path.isdir(docker_image_or_directory):
             docker_image_or_directory = os.path.abspath(docker_image_or_directory)
-            self.logger.info(
-                "Building a docker image from directory {}...".format(docker_image_or_directory)
+            docker_build_msg = "Building a docker image from directory: {}...".format(
+                docker_image_or_directory
             )
+            self.logger.info(docker_build_msg)
             self.logger.info("This may take some time")
             try:
-                image, _ = client.images.build(path=docker_image_or_directory)
-                ret_docker_image = image.id
-            except docker.errors.BuildError as e:
-                self.logger.error("Hey something went wrong with image build!")
-                for line in e.build_log:
+                # Set image tag to the dirname of the docker context.
+                # This may be confusing if try to build different images from different contexts,
+                # with the same folder name:
+                # /path1/my_env
+                # /path2/my_env
+                #
+                # If image with the tag `my_env` exists, it will be untagged.
+                tag = os.path.basename(docker_image_or_directory)
+                client_docker_low_level = docker.APIClient()
+                spinner = Spinner(docker_build_msg + "  ")
+                json_lines = []
+                # Build docker, rotate spinner according to build progress
+                # and save status messages from docker build.
+                for line in client_docker_low_level.build(
+                    path=docker_image_or_directory, rm=True, tag=tag
+                ):
+                    json_lines.append(json.loads(line.decode("utf-8").strip()))
+                    spinner.next()
+                # skip a line after spinner
+                print()
+
+                image_id = None
+                build_error = False
+                for line in json_lines:
+                    if "error" in line:
+                        build_error = True
+                        break
                     if "stream" in line:
-                        self.logger.error(line["stream"].strip())
+                        match = re.search(
+                            r"(^Successfully built |sha256:)([0-9a-f]+)$", line["stream"]
+                        )
+                        if match:
+                            image_id = match.group(2)
+                if image_id is None or build_error:
```

Comment:
probably best to have different errors here.

Reply:
I took this code snippet from:
https://github.com/docker/docker-py/blob/e6689e0bb9af849bd9d1509fd3b2db52e0d6d776/docker/models/images.py#L285-L297

As you can see they only rely on `Successfully built`, so in our case - if there is no image id just show all the messages and throw,